### ISSUE TITLE
Difficulty calculations with HHVM

### ISSUE THREAD

**Issue by aztech-dev** [afk11/bitcoin#12]
Currently tests fail with HHVM due to the difficulty tests.

It highlights a broader issue with your implementation, as current code is relying on built-in floats. Those do have a known, well-documented precision issue (it's both platform and installation dependent, as there is an INI setting to modify precision of floats).

A better solution would be to use for example bcmath (yeah...) with fixed precision scaling. However, this will require providing a userland implementation of log... 

It's definitely feasible, but maybe there is already a package providing that, I havent really checked.

An alternate route (which I don't favor) would be to dynamically modify the INI setting for precision when entering the diff function, and restoring it before exit.

**Comment by afk11** [comment]
I think that implementation was based on of the 6 lines of python here: https://en.bitcoin.it/wiki/Difficulty

It's essentially a shortcut to this which Gmp can't handle: 
```0x00000000FFFF0000000000000000000000000000000000000000000000000000 /
0x00000000000404CB000000000000000000000000000000000000000000000000 
= 16307.420938523983 (bdiff)```

BcMath would probably work fine with fixed precision though

**Comment by aztech-dev** [comment]
Oh stupid me. Of course.

It's also do-able with GMP then, just gotta do ($diff_1_target * (10 ** $precision)) / $current_target. That returns of course an int, and all that's left to do is insert the comma at the right place :)

I'll P/R that with GMP (avoids re-introducing a dependency to bcmath)

**Comment by afk11** [comment]
That would be cool, awesome! Yeah neither floats or bcmath are nice, glad we can do it through GMP.

**Comment by aztech-dev** [comment]
"Is" not "would" :p its in #13

**Comment by afk11** [comment]
Sweet, merged #13 so I'll close this!
